bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/scheduler/AbstractExpression.java

```diff
@@ -207,22 +207,51 @@ protected void prune() {
 
     @Override
     public Date getTimeAfter(Date afterTime) {
+
+        Date currentStartDate = getStartDate();
+
+        if (hasFloatingStartDate()) {
+            try {
+                clearCandidates();
+                setStartDate(afterTime);
+            } catch (IllegalArgumentException | ParseException e) {
+                logger.error("An exception occurred while setting the start date : '{}'", e.getMessage());
+            }
+        }
+
         if (getCandidates().isEmpty()) {
             try {
                 setStartDate(afterTime);
-                parseExpression(expression);
             } catch (ParseException e) {
-                logger.error("An exception occurred while parsing the expression : '{}'", e.getMessage());
+                logger.error("An exception occurred while setting the start date : '{}'", e.getMessage());
             }
         }
 
         if (!getCandidates().isEmpty()) {
-
-            Collections.sort(getCandidates());
-
-            for (Date candidate : getCandidates()) {
-                if (candidate.after(afterTime)) {
-                    return candidate;
+            if (getCandidates().size() == 1) {
+                return getCandidates().get(0);
+            } else {
+                while (getCandidates().size() > 1) {
```

Comment:
Why do you use > 1 here?
If this loop is iterated multiple times, isn't it valid, that the candidates consist of exactly one element?
At the end of the iteration you call `setStartDate(newStartDate);` I assume this could result in a candidates list of the minimum size (1, now).
Shouldn't it be `!getCandidates().isEmpty()`?